packages/styled-base/types/tests.tsx

```diff
@@ -0,0 +1,278 @@
+import * as React from 'react';
+import styled from '@emotion/styled-base';
+
+// tslint:disable-next-line: interface-over-type-literal
+type ReactClassProps0 = {
+  readonly column: boolean;
+};
+declare class ReactClassComponent0 extends React.Component<ReactClassProps0> {}
+
+interface ReactClassProps1 {
+  readonly value: string;
+}
+declare class ReactClassComponent1 extends React.Component<ReactClassProps1> {}
+
+interface ReactClassProps2 {
+  readonly column: number;
+}
+declare class ReactClassComponent2 extends React.Component<ReactClassProps2> {}
+
+// tslint:disable-next-line: interface-over-type-literal
+type ReactSFCProps0 = {
+  readonly column: boolean;
+};
+declare const ReactSFC0: React.SFC<ReactSFCProps0>;
+
+interface ReactSFCProps1 {
+  readonly value: string;
+}
+declare const ReactSFC1: React.SFC<ReactSFCProps1>;
+
+interface ReactSFCProps2 {
+  readonly value: number;
+}
+declare const ReactSFC2: React.SFC<ReactSFCProps2>;
+
+const Button0 = styled('button')`
+  color: blue;
+`;
+const Button1 = styled('button')({
+  color: 'blue',
+});
+
+(
+  <div>
+    <Button0 />
+    <Button0 type='button' />
+  </div>
+);
+(
+  <div>
+    <Button1 />
+    <Button1 type='button' />
+  </div>
+);
+
+const Input0 = styled('input', {
+  label: 'mystyle',
+})`
+  padding: 4px;
+`;
+const Input1 = styled('input', {
+  label: 'mystyle',
+})({
+  padding: '4px',
+});
+
+(
+  <Input0 />
+);
+(
+  <Input1 />
+);
+const Input2 = Button0.withComponent('input');
+
+const Canvas0 = styled('canvas', {
+  shouldForwardProp(propName) {
+    return propName === 'width' || propName === 'height';
+  }
+})`
+  width: 200px;
+`;
+const Canvas1 = styled('canvas', {
+  shouldForwardProp(propName) {
+    return propName === 'width' || propName === 'height';
+  }
+})({
+  width: '200px',
+});
+
+(
+  <Canvas0 />
+);
+(
+  <Canvas1 />
+);
+
+interface PrimaryProps {
+  readonly primary: string;
+}
+const Button2 = styled<'button', PrimaryProps>('button')`
```

Comment:
Why does `'button'` have to be duplicated?

Reply:
Yes, that's what I mean, could you add a comment explaining that for future reference?

Reply:
@mitchellhamilton Of course. How about adding a comment in the typing itself? (where the type parameter is defined)

Reply:
Both places would be great because people might look at the tests and type their stuff based on it.